---
title: "Exploratory data analysis with Yeast RNAseq data"
author: Taylor Reiter and Marian L. Schmidt
date: July 5th, 2017
output:
  html_document:
    theme: readable
    toc: true
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 2
bibliography: Bibliography/bibliography.bib
csl: Bibliography/plos-one.csl
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      message = FALSE,
                      engine = "R", # Chunks will always have R code, unless noted
                      error = TRUE,
                      fig.path="Figures/",  # Set the figure options
                      fig.align = "center")
```

## Data

This tutorial was written using data from Schurch et al.[@Schurch2016], where WT and MUT yeast were compared to test the underlying distribution of RNA-seq data, and to test which differential expression programs work best. They had 672 samples, we'll use 6 of them :) 

This markdown file was taken (and modified) from the [ANGUS 2017 lesson material](https://angus.readthedocs.io/en/2017/rmarkdown_rnaseq.html)

## Prepare for analysis

```{r install-packages, include=FALSE}
# Install all of the necessary packages
#install.packages("ggplot2")
#install.packages("RColorBrewer")
#install.packages("edgeR")
#install.packages("gplots")
#install.packages("GGally")
```
## Sniff Test of Data

We will use a pairwise comparison of the Salmon Transcripts Per Million (TPM) to understand how well our replicates replicated each other. 

We used a modified version of [gather-counts.py](https://github.com/ngs-docs/angus/blob/2017/_static/gather-counts.py) to gather TPM data (transcripts per million) in order to make a plot of these data.The modified script output a bunch of `.tpm` files with the following columns: 
| transcript | count |
|---|---|

We will use these data below.   

```{r import-tpm-files, include=FALSE}
# load package for edgeR to be able to make a DGE object
library("edgeR")
# Create an object with the tpm file names
files_tpm<-c("ERR458493.fastq.gz.quant.tpm",
             "ERR458494.fastq.gz.quant.tpm",
             "ERR458495.fastq.gz.quant.tpm",
             "ERR458500.fastq.gz.quant.tpm",
             "ERR458501.fastq.gz.quant.tpm",
             "ERR458502.fastq.gz.quant.tpm")

# Create a DGE object from the TPM data. The DGEList contains many data frames that are contained within an object. After reading in the data, we have $samples and we have $counts. We will begin working with the counts data. 
tpm<-readDGE(files_tpm)
```

In this report we will use a ggplot2-esque package to perform "sniff tests" to make pairwise comparison plots to see how well the replicates actually replicate each other.

```{r tpm-df, include=FALSE}
# ggplot does not know how to deal with objects of class matrix, so we will convert the tpm data to a dataframe first.
tpm_df<-as.data.frame(tpm$counts)
```
```{r head-tpm-df, include=TRUE}
head(tpm_df)
```

The first pairwise plot we make displays the wildtype samples. These are highly correlated, and thus replicate eachother quite well.

```{r pairwise-wt-plot-1, include=FALSE}
# create a plot using the first three columns of tpm data frame (the WT samples). 
library(GGally)
library(ggplot2)

pairwise_wt <- ggpairs(tpm_df[,1:3], columnLabels = c("ERR458493", "ERR458494", "ERR458495"), axisLabels = "internal", title = "WT pairwise comparisons") + theme_minimal()
```

```{r pairwise-wt-plot-2, include=TRUE}
pairwise_wt
```

We can do the same thing for the mutant samples. These are highly correlated as well, and also replicate each other well.

```{r pairwise-mut-plot-1, include=FALSE}
  # now create a plot using the last three columns of the tpm data frame (the MUT samples)
library(GGally)
library(ggplot2)
pairwise_mut <- ggpairs(tpm_df[,4:6], axisLabels = "internal", columnLabels = c("ERR458500", "ERR458501", "ERR458502"), title = "MUT pairwise comparisons") + theme_minimal()
```

```{r pairwise-mut-plot-2, include = TRUE}
pairwise_mut
```
These plots show us that our data look good -- it is valid to continue on and do differential expression with these data, because the replicates are true biological replicates.

Questions:
1. What species do you work with? Do you expect your biological replicates to be this similar?
2. Would you expect laboratory or environmental samples to be more variable?
3. What would alarming data look like?
4. How is TPM data different than raw counts? How is it different from edgeR or DeSEQ2 corrected results?

## Differential Expression

Since our data appears to be of high quality, let's perform differential expression with edgeR like we did before. We need to read in the proper count data, because edgeR expects raw counts, not counts that have been corrected in anyway. We will allow edgeR to perform correction for library size, and then estimate [dispersion](https://doi.org/10.1371/journal.pone.0081415).

```{r import-count-data, include=FALSE}
files <- c(
      "ERR458493.fastq.gz.quant.counts",
      "ERR458494.fastq.gz.quant.counts",
      "ERR458495.fastq.gz.quant.counts",
      "ERR458500.fastq.gz.quant.counts",
      "ERR458501.fastq.gz.quant.counts",
      "ERR458502.fastq.gz.quant.counts"
)
      
# Make a character vector named labels that will label the 6 samples above. 
# Note these are in the same order as the files -- WT are the first three, and MUT are the last three.
# They will be used later for the MDS plot
labels=c("WT_1", "WT_2", "WT_3", "MUT_1", "MUT_2", "MUT_3")

# Read in the data
data <- readDGE(files)

# check that the data are correct
print(data)
    
# We then need to describe the group that each sample is a part of, and estimate dispersion

# Create a vector to assign the "group" to the DGE object. This tells edgeR what the treatments here. 
# We have 6 samples and two treatments (3 WT, and 3 MUT).
group <- c(rep("wt", 3), rep("mut", 3))
# create a DGEList with the group variable
dge <- DGEList(counts=data, group=group)
dge<-calcNormFactors(dge)
# Make dispersion estimates. 
dge <- estimateCommonDisp(dge)
dge <- estimateTagwiseDisp(dge)
```

This will give us a DGE object. We can look at the DGE object and see some of the magic of edgeR.

```{r dge-object, include=TRUE}
head(dge$samples)
head(dge$counts)
```

Lastly, we perform differential expression!

```{r differential-expression, include=FALSE}
# perform the exact test one the DGE object for differential expression
et <- exactTest(dge, pair=c("wt", "mut"))

# grab the top 100,000 tags (n sets the number of tags that you want, the idea for using 100,000 is that all will be returned)
etp <- topTags(et, n=100000)
# let's take a look at the differentially expressed genes
etp # The first thing that is printed show "Comparison of groups:  mut-wt"
```
## Data Visualization

Differential expression outputs a large file, which may leave you thinking...what next?? 

### MDS Plot
We can make an MDS plot to understand how our samples relate to one another. From the edgeR manual: "a multi-dimensional scaling plot of the RNA samples in which
distances correspond to leading log-fold-changes between each pair of RNA samples. The leading log-fold-change is the average (root-mean-square) of the largest absolute log-foldchanges between each pair of samples. This plot can be viewed as a type of unsupervised clustering."
Our MUT samples cluster more closely to each other than to the WT samples. Does this agree with the sniff tests from pairwise comparisons we performed above?

```{r mds-plot, include=TRUE}
plotMDS(dge, labels=labels)
```

### MA Plot
We can also make an MA plot. From [wikipedia](https://en.wikipedia.org/wiki/MA_plot): "an MA plot is an application of a Blandâ€“Altman plot for visual representation of genomic data. The plot visualises the differences between measurements taken in two samples, by transforming the data onto M (log ratio) and A (mean average) scales, then plotting these values."
```{r ma-plot, include=TRUE}
etp$table$logFC = -etp$table$logFC # because our comparison is mut vs. wt, let's flip the tags for our plot
plot(etp$table$logCPM, etp$table$logFC, xlim=c(-3, 20), ylim=c(-12, 12), pch=20, cex=.3,
        col = ifelse( etp$table$FDR < .2, "red", "black" ) )
```
### Histogram
We can also make a histogram of the corrected p value (FDR). This shows that almost an equal number of differentially expressed and non-differentially expressed genes. 

```{r fdr-hist, include=TRUE}
library(ggplot2)
ggplot(etp$table, aes(FDR)) + geom_histogram(binwidth = .01) + ggtitle("Histogram of FDR for all Differentially Expressed Genes")
```

### Heatmaps
Heatmaps are also a very popular way to visualize data. Let's make a heat map of the 75 most variable genes!

```{r prep-var-heamap, include=FALSE}
# modified from http://combine-australia.github.io/RNAseq-R/06-rnaseq-day1.html
# make a dataframe of the counts that were normalized by edgeR
norm_counts<-dge$counts

# We estimate the variance for each row in the logcounts matrix
var_genes <- apply(norm_counts, 1, var)
head(var_genes)

# Get the gene names for the top 75 most variable genes
select_var <- names(sort(var_genes, decreasing=TRUE))[1:75]
head(select_var)
    
# Subset the normalized counts dataframe, grabbing the select_var genes
highly_variable_norm_counts <- norm_counts[select_var,]
    
# check the dimensions and the first few rows
dim(highly_variable_norm_counts)
head(highly_variable_norm_counts)
 
# Prepare for the heatmap
library(RColorBrewer) 

# Get some nicer colours
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
    
    
```
```{r heatmap-var, include=TRUE}
# Load library for heatmap
library(gplots)
heatmap.2(highly_variable_norm_counts, col=rev(morecols(50)), trace="none", 
              main="Top 75 most variable genes across samples", scale="row",
              labCol = c("WT", "WT", "WT", "MUT", "MUT", "MUT"))
```

Let's do the same thing again, but this time we will plot the top 75 least variable genes. Even though we did cpm normalization, a lot of these genes are still very lowly expressed. 

```{r prep-nonvar-heatmap, include=FALSE}
# First, let's get rid of the entries that are not expressed acrossed the board
keep_counts<-rowSums(cpm(dge)>1) >=2
dge_expressed <- dge[keep_counts, , keep.lib.sizes=FALSE]
norm_counts_expressed<-dge_expressed$counts
      
# We estimate the variance for each row in the logcounts matrix
var_genes_expressed <- apply(norm_counts_expressed, 1, var)
      
# Get the gene names for the top 75 most variable genes
select_var_least<- names(sort(var_genes_expressed, decreasing=FALSE))[1:75]
head(select_var_least)
    
# Subset the normalized counts dataframe, grabbing the select_var genes
not_variable_norm_counts <- norm_counts_expressed[select_var_least,]
    
# check the dimensions and the first few rows
dim(not_variable_norm_counts)
head(not_variable_norm_counts)
```

This is important to keep in mind when interpretting what this heatmap means.

```{r heatmap-non-var, include=TRUE}
# Load library for heatmap
library(gplots)
heatmap.2(not_variable_norm_counts, col=rev(morecols(50)), trace="none", 
              main="Top 75 least variable genes across samples", scale="row",
              labCol = c("WT", "WT", "WT", "MUT", "MUT", "MUT"))
```    

## References
